src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs

```diff
@@ -40,5 +40,54 @@ public static unsafe bool TryGetRawMetadata(this Assembly assembly, out byte* bl
 
             return InternalTryGetRawMetadata(new QCallAssembly(ref rtAsm), ref blob, ref length);
         }
+
+        [DllImport(RuntimeHelpers.QCall)]
+        private static extern unsafe int ApplyHotReloadUpdate(QCallAssembly assembly, byte* metadataDelta, int metadataDeltaLength, byte* ilDelta, int ilDeltaLength, byte* pdbDelta, int pdbDeltaLength);
+
+        /// <summary>
+        /// Hot reload update API
+        /// Applies an update to the given assembly using the metadata, IL and PDB deltas. Currently executing
+        /// methods will continue to use the existing IL. New executions of modified methods will use the new
+        /// IL. The supported changes are runtime specific - different .NET runtimes may have different
+        /// limitations. The runtime makes no guarantees if the delta includes unsupported changes.
+        /// </summary>
+        /// <param name="assembly">The assembly to update</param>
+        /// <param name="metadataDelta">The metadata changes</param>
+        /// <param name="ilDelta">The IL changes</param>
+        /// <param name="pdbDelta">The PDB changes. Current not supported on .NET Core</param>
+        /// <exception cref="ArgumentNullException">if assembly parameter is null</exception>
+        /// <exception cref="NotSupportedException">update failed</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta = default)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            RuntimeAssembly? runtimeAssembly = assembly as RuntimeAssembly;
+            if (runtimeAssembly == null)
+            {
+                throw new ArgumentException("Not a RuntimeAssembly", nameof(assembly));
```

Comment:
Messages should be in the .resx

Reply:
`throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);`

Reply:
Done.